Ambit3 with newer firmware

jadahl opened the pull request:
This pull request adds support for Ambit 3 devices with firmware version >= 2.0.4. It is based on the patch written by @tambit. What I did was clean it up and rewrite some parts, including changing it to keep backward compatibility with older firmwares.

Note that I have not tested with any other device than an Ambit3 Peak with the 2.0.4 firmware. There is one part of the code that is questionable (see device_info_get() where it has different commands for getting the firmware version, depending on the firmware version) and to actually be able to fix that, I will need synchronize logs for more devices with different firmware versions.

It also contains the ambit3-settings pull request (https://github.com/openambitproject/openambit/pull/108). Not sure I can make a new pull request based on a non-merged pull request, so please consider that patch not part of this series).

natanojl commented on src/libambit/device_driver_ambit3.c:
```diff
@@ -95,6 +102,25 @@ ambit_device_driver_t ambit_device_driver_ambit3 = {
+    struct generation *generation;
+
+    ARRAY_FOR_EACH(generations, generation) {
+        if (libambit_fw_version_number(generation->fw_version) >= libambit_fw_version_number(device_info->fw_version))
```
This comparison looks wrong. It should be <=, right?

jadahl replied:
True. Fixed locally so far.

natanojl replied:
Syncing moves seems to work with this change with my Ambit 3 Sport with firmware 1.0.16.

kbck commented:
Hi!
Tried to build your branch on Ubuntu 14.04 and got the following error:
```
------building libambit------
-- The C compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Found UDev: /usr/include  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kim/test/openambit/libambit-build
Scanning dependencies of target ambit
[  7%] Building C object CMakeFiles/ambit.dir/crc16.c.o
[ 14%] Building C object CMakeFiles/ambit.dir/debug.c.o
[ 21%] Building C object CMakeFiles/ambit.dir/device_driver_ambit.c.o
[ 28%] Building C object CMakeFiles/ambit.dir/device_driver_ambit3.c.o
In file included from /home/kim/test/openambit/src/libambit/sbem0102.h:29:0,
                 from /home/kim/test/openambit/src/libambit/device_driver_ambit3.c:29:
/home/kim/test/openambit/src/libambit/device_driver_ambit3.c: In function ‘get_ambit3_fw_gen’:
/home/kim/test/openambit/src/libambit/utils.h:114:5: error: ‘for’ loop initial declarations are only allowed in C99 mode
     for (size_t _i = 0; _i < ARRAY_LENGTH(_arr) && (_elem = &_arr[_i]); _i++)
```

BR,
Kim

jadahl commented:
Kim,

Does applying this fix the error:
```
diff --git a/src/libambit/CMakeLists.txt b/src/libambit/CMakeLists.txt
index 93b5042..0b4b11d 100644
--- a/src/libambit/CMakeLists.txt
+++ b/src/libambit/CMakeLists.txt
@@ -37,6 +37,8 @@ add_library (
   ${HIDAPI_SOURCE_FILES}
 )
 
+set_property(TARGET ambit PROPERTY C_STANDARD 99)
+
 target_link_libraries(
   ambit
   ${HIDAPI_LIBS}
```

kbck commented:
Hi!

yes it works, now I'm only getting a warning message:
```

 [ 28%] Building C object CMakeFiles/ambit.dir/device_driver_ambit3.c.o
/home/test/openambit/src/libambit/device_driver_ambit3.c: In function ‘log_read’:
/home/test/openambit/src/libambit/device_driver_ambit3.c:195:14: warning: variable ‘log_entries_notsynced’ set but not used [-Wunused-but-set-variable]
     uint16_t log_entries_notsynced = 0;
```

Is it ok to enable C99? Minimum CMake version seems to be:
```
cmake_minimum_required (VERSION 2.8.5)
```
Is this supported by 2.8.5?

jadahl commented:
> Is it ok to enable C99? Minimum CMake version seems to be:

That'll have to be up to the maintainers. If not I'd have to open code the ARRAY_FOR_EACH since such a macro may not be implemented without C99 support.

As for the warning, I get that too (except on other lines), and would have to fix those indeed.

kbck commented on src/libambit/libambit.c:
```diff
@@ -320,13 +321,27 @@ void libambit_log_entry_free(ambit_log_entry_t *log_entry)
     LOG_INFO("Reading device info");
 
-    if (libambit_protocol_command(object, ambit_command_device_info, komposti_version, sizeof(komposti_version), &reply_data, &replylen, 1) == 0) {
+    switch (info->product_id) {
```
Could this kind of information be in device_support.c? Adding support for new devices would otherwise require changes here & maybe more places, I'm not that familiar with this code yet.
An example: I've got a Ambit3 Vertical (product id 0x002c) .. which doesn't work with openambit.. 
Note, product id etc for Ambit3 Vertical can be found here: https://github.com/kbck/openambit/tree/ambit3_vertical

jadahl commented:
I suppose detection could be moved around a bit. Does your Ambit3 Vertical work if you add that patch and the 0x002c in that extra place?

kbck commented:
Added 0x002c to the switch statement, but it doesn't work ..
Log:
```
libambit INFO: ambit_device_info_new(): HID  : /dev/hidraw0: 'Suunto Ambit' (serial: 8820125121000000, VID/PID: 1493/002c)
libambit INFO: device_info_get(): Reading device info
libambit INFO: ambit_device_info_new(): preferring known name over HID 'Suunto Ambit'
libambit INFO: ambit_device_info_new(): Ambit: /dev/hidraw0: 'Suunto Ambit3 Vertical' (serial: 8820125121000000, VID/PID: 1493/002c, nick: Kaka, F/W: 1.0.27, H/W: 71.2.0, supported: YES)
Device: Suunto Ambit3 Vertical, serial: 8820125121000000
F/W version: 1.0.27
libambit INFO: libambit_device_driver_status_get(): Reading device status
Current charge: 79%
libambit INFO: personal_settings_get(): Reading personal settings
libambit INFO: log_read(): Reading log headers
libambit WARNING: get_memory_maps(): Failed to read memory map key
libambit INFO: libambit_close(): Closing
```
Haven't looked at this before, so I don't know how it works ..

Regards

jadahl commented:
If you add the 0x002c case next to the default case (so that you use the default "komposti_version"), does it work then?

You don't happen to have a usbcap of successfully synchronizing on Windows have you?

kbck commented:
Hi!
I haven't had any time to answer your question yet (lots of work and other stuff). But l'll get back to you soon and send you a pcap file.

kbck commented:
Hi!
tested with the default "komposti_version", that didn't work either...

I've mailed you a pcap file.

Regards

kbck commented on wireshark_dissector/ambit-dissector.c:
```diff
@@ -32,6 +32,9 @@ typedef struct ambit_protocol_type {
 } ambit_protocol_type_t;
 
+static gint fw_version = 0;
+statig gint fw_verson_2_0_0 = 2 << 16;
```
Typos?

microwan commented:
Hello Jadahl ,

I have tested your Pull Request
It is ok. I can collect  the logs on Ambit 3 Peak on firmware 2.0.45
but Qt GUI the watch is not supported
Maybe the id is not correct.

gbeste commented:
Hi!

Does openambit work with Ambit3 Peak, FW version 2.2.16? I cannot get mine to work...

Idiotein commented:
Hi jadahl, got around to trying again with my Peak firmware 2.2.16, it doesn't work.
Only message I get is "libambit WARNING: get_memory_maps(): Failed to read memory map key", and no debug log is generated (despite building with "-DCMAKE_BUILD_TYPE=Debug").

Also I caught a typo in "wireshark/ambit-dissector.c", line 35 for me:
`typedef struct ambit_protocol_type {
    guint32 command;
    char *name;
    gint (*dissector)(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, void *data _U_);
} ambit_protocol_type_t;

static gint fw_version = 0;
-static gint fw_verson_2_0_0 = 2 << 16;
+static gint fw_version_2_0_0 = 2 << 16;`

It's "version" instead of "verson", with this change it builds fine here.

I'll try to make captures on windows later today, will report here.

Thanks.

Idiotein commented:
Ok, late to the party, I see @kbck already reported the typo, ignore the noise.
The wireshark dissector doesn't load here, wireshark is 2.2.1, I get "undefined symbol new_create_dissector_handle".
Anyway, here come the captures. Let me know if they are usable, I am using a new capture setup.
Thanks.
[Suunto.zip](https://github.com/openambitproject/openambit/files/575126/Suunto.zip)

christophe-lyon commented:
Hi, I'm a newbie here. I've got an ambit3 peak with FW 2.2.16. I've tried this branch with the typos fixed and it still does not work.
The messages I have are:
libambit INFO: ambit_device_info_new(): ignoring unknown device (VID/PID: 03f0/034a)
libambit INFO: ambit_device_info_new(): ignoring unknown device (VID/PID: 03f0/034a)
libambit INFO: ambit_device_info_new(): ignoring unknown device (VID/PID: 0458/0007)
libambit INFO: libambit_device_driver_status_get(): Reading device status
libambit INFO: personal_settings_get(): Reading personal settings
libambit INFO: libambit_device_driver_date_time_set(): Writing date and time to clock
libambit INFO: log_read(): Reading log headers
libambit WARNING: get_memory_maps(): Failed to read memory map key
libambit INFO: libambit_close(): Closing

AndreaGiardini commented:
Hello everyone

Any update on this pull request?

obucek commented:
hello to all.
I install Openambit version 0.3.
![izbor_001](https://cloud.githubusercontent.com/assets/12692425/22854909/1e00f014-f077-11e6-9f75-a8fe82827df1.png)

Can you help me to solve problem. I have linux (Deepin) and dont know how to configure and install it from source.
Thx in advanced and all the best

Oliver

jadahl commented:
I updated the branch and fixed some things. What currently doesn't seem to work is getting heart rate log data, at least from the default "Running" activity, my watch does not include any such log data. FWIW, they are available when syncing via the Android app, so it's probably something i'm not doing when talking via USB. If anyone have any USB logs to share with me when synchronizing different types of activities on different firmwares with the official desktop application, I would gladly take a look at them.

Locally I have a patch that changes the 'komposti-version', which I didn't include. It might make things work on even newer firmwares, but I'm not sure. If you dare to try, apply this patch:

`diff --git a/src/libambit/libambit.c b/src/libambit/libambit.c
index ed79bf1..f9a90c4 100644
--- a/src/libambit/libambit.c
+++ b/src/libambit/libambit.c
@@ -53,7 +53,7 @@ static ambit_device_info_t * ambit_device_info_new(const struct hid_device_info
  * Static variables
  */
 static uint8_t komposti_version_default[] = { 0x02, 0x00, 0x2d, 0x00 };
-static uint8_t komposti_version_ambit3[] = { 0x02, 0x03, 0x06, 0x00 };
+static uint8_t komposti_version_ambit3[] = { 0x02, 0x04, 0x59, 0x00 };
 
 /*
  * Public functions`

manisandro commented:
Just for the record, if this is deemed stable I'm happy to merge this. I don't have an Ambit 3 to test or particular knowledge of the protocol, but I'm interested in keeping the project alive.

jadahl commented:
FWIW, ambit3 support is still incomplete, but at least on my watch, it doesn't work at all without this branch (and the other pull request with the settings patches). I don't have a windows computer/install, so the most major thing stopping me from making any significant progress right now is lack of logs from successful synchronizations.

So, anyone wanting to help out, the easiest way to help is providing usb synchronization logs.

eppesuig commented:
Hello Jadahl, I may collect a log using windows on virtualbox. Do I have to use an usb sniffer on Linux or Windows? Is there any specific option you need me to use?

jadahl commented:
@eppesuig not sure whether to to listen on the host or the guest actually. Synchronization of logs with different types of activity types would be helpful. I have never used the windows app so i don't know if there are any options that are relevant.

AndreaGiardini commented:
Is there a document that we can read with instructions to provide you meaningful logs?

jadahl commented:
There is the FAQ at http://openambit.org/wireshark-dissector/

Idiotein commented:
Hello all, here comes a sync with a few activities (swim, trail, run and maybe indoor). I tried to trick movescount app to re-download all logs from watch by creating a fresh account, let me know if it didn't work so I can add new captures.
It's from Ambit3 Peak with latest firmware.

@eppesuig and @AndreaGiardini , I did captures both from linux host (Debian) and inside Virtualbox with Wireshark, both work equally well. Latest Wireshark for Windows includes usb capture utility (usbpcap) so nothing else needed to capture from the windows guest or bare metal install.
If the Windows account isn't admin usb may not show up in Wireshark, in this scenario capturing from Linux host is easier.
[Ambit3Peak_2017-03.pcapng.gz](https://github.com/openambitproject/openambit/files/821002/Ambit3Peak_2017-03.pcapng.gz)

Idiotein commented:
Hi @jadahl,

tried the build ambit3 with the "komposti" patch (Debian Sid amd64), builds fine. My watch (Ambit3 Peak) starts beeping early with "sync with moveslink" error, Openambit gui gives me all green anyway but ultimately errors with "sync failed". Sure enough there isn't any log in the .openambit folder.
I tried to capture with Wireshark, tell me if it's any use.
[OpenAmbit3_failedsync_Ambit3Peak_201703-2.pcapng.gz](https://github.com/openambitproject/openambit/files/821149/OpenAmbit3_failedsync_Ambit3Peak_201703-2.pcapng.gz)
[OpenAmbit3_failedsync_Ambit3Peak_201703.pcapng.gz](https://github.com/openambitproject/openambit/files/821150/OpenAmbit3_failedsync_Ambit3Peak_201703.pcapng.gz)

At least the watch is not bricked, and I didn't lose anything on it at first inspection.

eppesuig commented:
Hi @jadahl, this is my dump from a quick synchronization with an Ambit 3 Peak. It only contains one track for sport "Bicicletta" that is (bicycle).
[ambit3pea-20170306k.zip](https://github.com/openambitproject/openambit/files/822635/ambit3pea-20170306k.zip)

Idiotein commented:
Some more data, one swim and one run.
[Ambit3Peak-13032017.zip](https://github.com/openambitproject/openambit/files/840355/Ambit3Peak-13032017.zip)

Gerold-R commented:
@Idiotein 

> tried the build ambit3 with the "komposti" patch (Debian Sid amd64), builds fine. My watch (Ambit3 Peak) starts beeping early with "sync with moveslink" error, Openambit gui gives me all green anyway but ultimately errors with "sync failed". Sure enough there isn't any log in the .openambit folder.
> I tried to capture with Wireshark, tell me if it's any use.
> OpenAmbit3_failedsync_Ambit3Peak_201703-2.pcapng.gz
> OpenAmbit3_failedsync_Ambit3Peak_201703.pcapng.gz

I had a quick look at your files above and it seems to me that you are using an old Komposti (K-Lib) Version number.

Your data send to the watch to get the device info is:
3F 18 5D 10 01 00 2F B8 00 00 01 00 00 00 00 00 04 00 00 00 02 03 06 00 BE D4
Komposti version in this command is  02 03 06 00 instead of 02 04 59 00

When this version number is too low, the watch will beep.

You also get an error when OpenAmbit sends the command "ambit_command_lock_check"
3F 14 5D 0C 01 00 2D 8E 0B 19 05 00 09 00 01 00 00 00 00 00 D9 D4
The reponse has the error flag set to 02: 
3F 14 5D 0C 01 00 2D 8E 0B 19 02 02 09 00 01 00 00 00 00 00 44 0E
This probably means (after my experience), that the command is not supported. Maybe cause of the wrong version number. 

I found 3 more commands send to the watch and the watch answered with an error.

Idiotein commented:
Thanks  @Gerold-R for looking into this. I rebuilt making sure proper "komposti" version was used, it no longer beeps but I still don't get any log in .openambit folder. It errors out with:
"libambit WARNING: libambit_pmem20_log_init(): Failed to read first data chunk
libambit ERROR: libambit_pmem20_log_read_entry_address(): Failed to parse log entry header correctly"

openambit also frequently crashes during log parsing. I will try to capture what is happening today.

jadahl commented:
@Idiotein don't expect it to work yet, and the crash is probably just because openambit seems to not have been prepared (programming wise) for not understanding input. There are almost no out-of-bounds checks for example.

From the logs you provided I've seen that commands are sent prior to log syncing, logs that my watch don't understand, and I expect these to be needed, as well as adaptations to the parsing.

There are these types of differences between generation 1 and generation 2 of firmwares too, as in there are more fields squeezed in into replies and the code has to be adapted to deal with this for the newest firmware generations.

landas commented:
Hi,

Is there been done more work to this code than commited to github?
I have merged this code (not wireshark_dissector/ambit-dissector.c) to work with current version of openambit.

https://github.com/landas/openambit/tree/jadahl
https://github.com/openambitproject/openambit/compare/master...landas:jadahl

jadahl commented:
@landas only a wip branch locally, which is mostly adding the unknown commands as seen in provided logs, but yet to update any parsing.

What stopped you from updating the dissector? The old version doesn't work at all on recent versions of wireshark.

landas commented:
@jadahl it took me to much time to figure out the merge conflicts. I would be happy if you could fix the merge conflict in wireshark_dissector/ambit-dissector.c and make it a seperate pull request :smile:

Idiotein commented:
@jadahl don't worry, I have been using libre software long enough not to expect it to work when I need it, or not break, or be delivered on time or follow the roadmap ;-) . More seriously, let me know what info I can provide that could be useful. I have a 3 Peak with latest firmware, an Ambit2 also, and access to windows systems. Tell me what kind of capture you need and I will do my best, I already posted some.
Thanks.

landas commented on src/libambit/device_driver_ambit3.c:
```diff
@@ -371,6 +588,14 @@ static int parse_log_header(const uint8_t *data, ambit3_log_header_t *log_header
+        break;
+      case 0x7a: /* gen2 fw */
+      case 0x8a: /* gen2 fw */
+        offset += (strlen(log_header->header.activity_name)+1);
```
@jadahl do you think this line will cause a problem with non-english characters?
Should it calculate the offset based on the ISO-format in  (data + offset) ?

Do you know why this offset exsist in 0x8a type?

jadahl replied:
That sounds like a possibility indeed. Should probably strlen() the input data, rather than the UTF-8 converted one, and use that to offset the offset.

landas commented:
Commits are in #146
